File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.go
func reverseWords(s string) string {
    start := 0
    end := len(s) -1
    var front, back []byte
    
    for start <= end {
    
        // skip spaces before a word
        for start <= end && s[start] == ' ' {
            start++
        }
        // skip spaces after word
        for start <= end && s[end] == ' ' {
            end--
        }

        // pick the first word and add it to front of back string
        var leftWord, rightWord []byte
        for start <= end &&  s[start] != ' ' {
            leftWord = append(leftWord, s[start])
            start++
        }
        if leftWord != nil {
            leftWord = append([]byte{' '}, leftWord...)
            back = append(leftWord, back...)
        }
        // pick the word from right and append it to end of front string
        for start <= end &&  s[end] != ' ' {
            rightWord = append([]byte{s[end]}, rightWord...)
            end--
        }
        if rightWord != nil {
            rightWord = append(rightWord, ' ')
            front =  append(front, rightWord...)
        }
        
        start++
        end--
    }
    return string(append(front, back[1:]...))
}

File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.cpp
class Solution {
public:
    string reverseWords(string s) {
        reverse(s.begin(), s.end());
        int ptr = 0;
        string ans = "";
        while(ptr < s.size()) {
            // skip whitespaces
            while(ptr < s.size() && s[ptr] == ' ') {
                ptr++;
            }
            if(ptr == s.size()) break;
            // Word found
            string word = "";
            while(ptr < s.size() && s[ptr] != ' ') {
                word = s[ptr] + word;
                ptr++;
            }
            ans += word + " ";
            ptr++;
        }
        // cout << ans << endl;
        return ans.substr(0, ans.size()-1);
    }
};